### PR TITLE
wip: Add valid_before and valid_after fields to transaction

### PR DESCRIPTION
**Pull request by mikayla-maki**
This PR aspires to implement https://github.com/penumbra-zone/penumbra/issues/1915

I have removed expiry height, and added 2 new fields for valid_before / valid_after.

TODO:
- [x] Figure out how to add a state machine test for this behavior
- [x] Remove expiry_height
- [x] Do due diligence on time representations
- [ ] Get a pass from the penumbra dev team on:
  - [ ] Wording in documentation / errors
  - [ ] Proto numbering
- [ ] Have the PR pass review :D

### CODE REVIEW

File: component/src/action_handler/transaction/stateful.rs [hunk 1]
```diff
@@ -32,3 +33,31 @@ pub(super) async fn fmd_parameters_valid<S: StateRead>(
         height,
     )
 }
+
+pub(super) async fn timestamps_are_valid<S: StateRead>(
```
Comment (mikayla-maki) [inline component/src/action_handler/transaction/stateful.rs:37]: I wasn't sure exactly where to add the checks for these fields, and what the difference between a panic and an `Err()` is in this context, this is my best guess. I also couldn't find any check for the expiry height anywhere, so I added it here.

File: proto/proto/penumbra/core/transaction/v1alpha1/transaction.proto [hunk 2]
```diff
@@ -41,6 +41,24 @@ message TransactionBody {
...
   optional bytes encrypted_memo = 6;
+  // A unix timestamp, in nanoseconds.
+  // If this value is set, the transaction is valid iff the most recent block header
+  // timestamp is < this value.
+  //
+  // If the value is zero, this value is considered unset.
+  // [TODO: Double check that this actually does behave like google's protobufs]
+  // The behavior of this time stamp should match google's protobuf specification:
+  // [specification]: https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp
+  uint64 valid_before = 7;
```
Comment (mikayla-maki) [inline proto/proto/penumbra/core/transaction/v1alpha1/transaction.proto:52]: This is supposed to match exactly the Timestamp type in IBC, which has this note on it. I don't have the context to verify this yet, hence the TODO.

File: proto/proto/penumbra/view/v1alpha1/view.proto [hunk 3]
```diff
@@ -106,6 +106,10 @@ message TransactionPlannerRequest {
     core.crypto.v1alpha1.Fee fee = 2;
     // The memo for the requested TransactionPlan
     string memo = 3;
+    // The valid_before timestamp for the requested TransactionPlan, formatted as an RFC 3339 timestamp
+    optional string valid_before = 4;
```
Comment (mikayla-maki) [inline proto/proto/penumbra/view/v1alpha1/view.proto:110]: From what I remember of the penumbra architecture, this is  the internal RPC call that converts from UI to internal penumbra data types. As this is more 'user facing', I thought a string would be nicer here.

File: view/src/service.rs [hunk 4]
```diff
@@ -354,6 +354,19 @@ impl ViewProtocolService for ViewService {
             )
             .expiry_height(prq.expiry_height);
 
+        if let Some(timestamp) = prq.valid_after {
+            let time = tendermint::Time::parse_from_rfc3339(timestamp.as_str()).map_err(|e| {
+                tonic::Status::invalid_argument(format!("Could not parse valid_after: {e:#}"))
+            })?;
+            planner.valid_after(time);
```
Comment (mikayla-maki) [inline view/src/service.rs:361]: Note that the view service does not depend on IBC, only tendermint, so this code has to use the tendermint `Time` type

File: transaction/src/plan.rs [hunk 5]
```diff
@@ -40,6 +41,8 @@ pub struct TransactionPlan {
     pub fee: Fee,
     pub clue_plans: Vec<CluePlan>,
     pub memo_plan: Option<MemoPlan>,
+    pub valid_before: Timestamp,
```
Comment (mikayla-maki) [inline transaction/src/plan.rs:44]: Note that the transaction crate does not depend on tendermint, only IBC, and so has to use the IBC `Timestamp` type.

**Comment by hdevalence** [comment]
This looks great!

The only thing I'm not sure about is modeling the timestamps using the IBC timestamp types -- I'm not sure we want to link our transaction domain types to the IBC types that way.  Using a Tendermint timestamp type might also be problematic, because IIRC there are a bunch of weird rules it implements to be exactly compatible with some accidental behavior of a Go library for timestamp formatting, and we might not want to pull in those consensus rules by accident.  That said, at the moment we don't have a timestamp type of our own, so we don't have a ready-to-go alternative.

We're currently in the process of adding time-based dynamic epochs (#2070), which will have to define our own timestamp type. So I think the cleanest thing to do would be:

- We will finish and merge work on #2070 that includes a timestamp type;
- We will rebase your branch on top of that work, and swap out the timestamp types;
- Then, pick up code review again.

Sorry for the mis-planning and the delay -- I know we'd listed it as a good first issue, and it turned out to be more intertwined than expected -- but I think if we wait a bit, we can rebase it on top of that work and then resume the process!

**Comment by mikayla-maki** [comment]
Sounds good to me! Guess we'll leave this hanging for a while, excited to see what y'all come up with :D

Also, glad to hear that the code I wrote is otherwise OK, this is a lot more of the system than I'm used to interacting with :)

**Comment by hdevalence** [comment]
We didn't land this :( 

Sorry to have let you down on it, but I think it's best that we close this out for now.

**Pull request closed without merging**